Common/Algorithm/Framework/Alphas/Analysis/InsightManager.cs

```diff
@@ -127,28 +142,31 @@ public void InitializeExtensionsForRange(DateTime start, DateTime end, DateTime
         /// <param name="generatedInsights">Any insight generated by the algorithm at the frontier time</param>
         public void Step(DateTime frontierTimeUtc, ReadOnlySecurityValuesCollection securityValuesCollection, GeneratedInsightsCollection generatedInsights)
         {
-            if (generatedInsights != null && generatedInsights.Insights.Count > 0)
+            lock (_lock)
```

Comment:
Just wondering if the lock surface can be reduced a little 🤔 
Maybe locking only L157 and L164 is enough?

Reply:
Thought about it too, what stopped me is that `extensions` are not thread safe either (see https://github.com/QuantConnect/Lean/issues/3155#issuecomment-492731892 for example), so I just decided to wrap everything since there wouldn't be any real difference